### PR TITLE
feat: add lru cache

### PR DESCRIPTION
**Pull request by andrewleap-optimizely**
Summary
-------

Add a generic LRU cache to support efficient caching of the ODP segments results. 

- Add Least Recently Used (LRU) cache
- design to allow developer to pass in different caching implementation. Provide default.

Test plan
---------
Added:
- lru_cache_spec.rb

Ticket:
---------
[OASIS-8435](https://optimizely.atlassian.net/browse/OASIS-8435)

### CODE REVIEW

File: lib/optimizely/odp/lru_cache.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+      @cache_mutex.synchronize do
+        return nil unless @map.include?(key)
+
+        old_element = @map[key]
+        @map.delete(key)
+        return nil if old_element.stale?(@timeout)
+
+        element = CacheElement.new(old_element.value)
+        @map[key] = element
```
Comment (jaeopt) [inline lib/optimizely/odp/lru_cache.rb:52]: It looks like ruby hash keeps insertion order for 1.9+. Min ruby version we support?

File: lib/optimizely/odp/lru_cache.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      return nil if @capacity <= 0
+
+      @cache_mutex.synchronize do
+        return nil unless @map.include?(key)
+
+        old_element = @map[key]
+        @map.delete(key)
+        return nil if old_element.stale?(@timeout)
+
+        element = CacheElement.new(old_element.value)
```
Comment (jaeopt) [inline lib/optimizely/odp/lru_cache.rb:51]: Timestamp is not supposed to change for lookup. This new creation will overwrite the old timestamp.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks great. A fix suggested.

File: lib/optimizely/odp/lru_cache.rb (hunk 1 above)
Comment (andrewleap-optimizely) [reply lib/optimizely/odp/lru_cache.rb:52]: That's one of the main reasons I pushed for updating the minimum supported ruby version. As of 4.0.0, we now only support >=2.7

Comment (jaeopt) [reply lib/optimizely/odp/lru_cache.rb:52]: @andrewleap-optimizely That min version info should be in README.md file and our docs. Updated?

Comment (andrewleap-optimizely) [reply lib/optimizely/odp/lru_cache.rb:52]: The minimum version is listed on the rubygems page but yea we should add it to the readme as well. Looks like a lot of the sdks are missing that information on their readme/docs, may be worth discussing.

File: lib/optimizely/odp/lru_cache.rb (hunk 2 above)
Comment (andrewleap-optimizely) [reply lib/optimizely/odp/lru_cache.rb:51]: good catch. Fixed and added test to confirm.

**Review by jaeopt — APPROVED** [review]
LGTM

**Review by zashraf1985 — APPROVED** [review]
Looks Great!

**Pull request merged**